src/ngraph_encapsulate_op.cc

```diff
@@ -520,16 +520,16 @@ class NGraphEncapsulateOp : public OpKernel {
   int m_ngraph_cluster;
   std::vector<bool> m_input_is_static;
 
-  static std::shared_ptr<ng::runtime::Backend> s_ng_backend
+  static std::weak_ptr<ng::runtime::Backend> s_ng_backend_wptr;
+  std::shared_ptr<ng::runtime::Backend> m_ng_backend
       GUARDED_BY(s_ng_backend_mutex);
```

Comment:
Just a note: I read up on GUARDED_BY and wondering if we should be locking `s_ng_backend_mutex` prior to tensor creation via `m_ng_backend->create_tensor`?
```
// GUARDED_BY allows the user to specify a particular mutex that should be
// held when accessing the annotated variable. 
```

Reply:
m_ng_backend will be copy to s_ng_backend_wptr; so the s_ng_backend_wptr, so it will be point to the same resource. That's why s_ng_backend_wptr need to be inside the mutex.

Reply:
Right, I'm wondering something a little different. `m_ng_backend` is `GUARDED_BY(s_ng_backend_mutex)`. The above comment indicates that a lock should be held whenever the annotated variable (`m_ng_backend` in this case) is accessed - for example, when `create_tensor` is called. My guess is that it isn't critical, but I'm not familiar enough with this TF macro to be sure, and so just wanted to point it out.

Reply:
to my understanding, `create_tensor` is thread safe. Since it does not modify non-local data.

Reply:
Yeah, since we default to device 0, it should be fine. Since we assume a MT environment in the TF bridge, we should probably add `cudaSetDevice(0)` in the GPU_Backend::create_tensor, or the GPU_TensorView to avoid future bugs when we add multi-gpu support and are creating tensors from different threads.

Regarding `GUARDED_BY`, from here: https://clang.llvm.org/docs/ThreadSafetyAnalysis.html it suggests that use of this without the lock will emit a compile-time warning -- but only for clang, I believe.